Add a new schema for Bears

surli opened the pull request:
Should fix #491

surli commented on resources/bears-example.json:
```diff
@@ -0,0 +1,127 @@
+    "failingModule": "flink-tests",
+    "failingClasses": [
+      {
+        "testClass": "org.apache.flink.util.FileUtilsTest",
```
Then it would look something like that, which I find far more difficult to read:

```json
{
        "testClass": "org.apache.flink.util.FileUtilsTest",
        "failingMethods": [
          {
            "testMethod": "testOne",
            "failures": {
                "name": "java.lang.AssertionError",
                "isError": false,
              }
          },
          {
            "testMethod": "testTwo",
            "failures": {
                "name": "java.lang.AssertionError",
                "isError": false,
              }
          },
          {
            "testMethod": "test3",
            "failures": {
              "name": "java.lang.AssertionError",
              "isError": false,
            }
          },
          {
            "testMethod": "test4",
            "failures": {
              "name": "java.lang.AssertionError",
              "isError": false,
            }
          },
          {
            "testMethod": "test5",
            "failures": {
              "name": "java.lang.AssertionError",
              "isError": false,
            }
          },
          {
            "testMethod": "test6",
            "failures": {
              "name": "java.lang.AssertionError",
              "isError": false,
            }
          }
          ,{
            "testMethod": "test7",
            "failures": {
              "name": "java.lang.AssertionError",
              "isError": false,
            }
          }
        ]
      }
```

fermadeiral replied:
And if we put only failingMethod but with fully-qualified name?

surli replied:
It's the same problem: a failing method as only one failure, so we'll get a very long list of failing methods with their failures. And if it's only assertion errors, we'll see AssertionErrors everywhere.

So the question here: do you think you'll need to analyze for each test methods the failure? Or do you need more general metrics about the kind and number of failures? I can also display a list of failing methods without their attached failure.

fermadeiral replied:
I can't say for you a real case I will need the detailed information, but I think it should be better to have it and to delete in the future if this is useless than to not have it and to be necessary.

To be honest, I don't think I will analyze the .json file directly (maybe only to check very small things quickly), but I will use bears-usage or the ongoing web interface instead.

surli replied:
ok then I propose a middle way: to provide in overallMetrics the list of failures with number of occurences on the whole test suite, which also define if the test was an error or not. And then a detailed list of testclasses and methods with the obtained failure. It would look like that:

```
overallMetrics: {
  failures: [
     {
         className: "java.lang.AssertionError",
         isError: false,
         occurences: 19
     },
     {
         className: "java.lang.NullPointerException",
         isError: true,
         occurences: 19
     }
  ]
},
metricsByClass: [
  {
     testClass: "org.acme.MyClass",
     failures: 1,
     errors: 1
  },
  {
     testClass: "org.acme.OtherClass",
     failures: 1,
     errors: 0
  }
],
failureDetails: [
  {
     testClass: "org.acme.MyClass",
     testMethod: "test1",
     failure: "java.lang.AssertionError"
  },
  {
     testClass: "org.acme.MyClass",
     testMethod: "test2",
     failure: "java.lang.NullPointerException"
  },
  {
     testClass: "org.acme.OtherClass",
     testMethod: "test1",
     failure: "java.lang.AssertionError"
  }
]
```

WDYT?

fermadeiral replied:
It seems good. But don't you think we also need "isError" in each item in "failureDetails" to be really complete?

surli replied:
I can put it yes. OK so let's go to this solution

surli commented:
Just tested with ajv (https://github.com/jessedc/ajv-cli) to check the validity of the example against the schema and it's valid.

surli commented:
@fermadeiral yes! sorry I forgot to change the status

fermadeiral commented:
> @fermadeiral yes! sorry I forgot to change the status

;)

Last question: do you really think it's a good idea for branches in Bears to have also the file "repairnator.json"? I'm asking this because sometimes I feel we will have redundant info, and I concern that in scripts and also in bears-usage tool I will need to search info in both files. I just want to be sure on what we are doing here, if we keep also "repairnator.json" for Bears or not, because if it's not a good idea to keep it I would ask you to add some info in Bears schema from repairnator schema.

surli commented:
> if it's not a good idea to keep it I would ask you to add some info in Bears schema from repairnator schema.

Consider you won't have both files: what extra info do you need?

surli commented:
ping @fermadeiral you did not answer my last question

fermadeiral commented:
> Consider you won't have both files: what extra info do you need?

`"StepsDurationsInSeconds": { ... }`: I think this is interesting because people can have an idea on which bugs are faster or slower to be reproduced, mainly because of the steps build and test.

fermadeiral commented:
Simon, I also would like to include `bears_version`, which should contain a string.

surli commented:
@fermadeiral I'll finish this one right now but you can start the implem even if the schema is not completely finished.

fermadeiral commented:
> @fermadeiral I'll finish this one right now but you can start the implem even if the schema is not completely finished.

Ok, Simon, I'm on it.

surli commented:
I'm a bit annoyed with the `"StepsDurationsInSeconds": { ... }`: if the steps are renamed the schema won't be good anymore. 
Are you sure it's mandatory and what should be its format then?

fermadeiral commented:
> I'm a bit annoyed with the "StepsDurationsInSeconds": { ... }: if the steps are renamed the schema won't be good anymore.
Are you sure it's mandatory and what should be its format then?

What about not putting it in the schema, but recording it in the json file anyway?

surli commented:
> What about not putting it in the schema, but recording it in the json file anyway?

I'm not really a huge fan. I prefer that we define clearly what should be interesting and then we define it in the schema: like, you want to know the duration for the part of the pipeline:
 - to clone,
 - to build,
 - to test
 - to repair

Then we compute those time maybe by aggregating data from several steps of the pipeline and we put them in the JSON.

fermadeiral commented:
@surli, I think the four steps you just described are the good ones. I mean, the info on how much time was spent to checkout, or to get test info, or to push, is not really interesting.

For Bears, I think it's nice to put two "build" and "test" as we have a pair of builds.

For Repairnator, I think it's nice to have the "repair" one by tool.

I know we are not mixing the two now, but I'm just expressing what I think because we might already prepare for the future some type of stuff (I mean, if we plan to build a two-floor house, but we are do a one-floor house now, we might make the structure for this house now stronger to be adapted later for another floor).

surli commented:
@fermadeiral so I updated the new schema with duration info: we have 4 global steps:
 - cloning,
 - building,
 - testing,
 - fixing

For each global step, you have to specify:
 - number of steps,
 - names of implicated steps
 - overall duration

WDYT?

fermadeiral commented:
@surli, it seems ok. To really make sure we are in the same page, please, check if the calculation of the building and testing global steps will be:

When `ProjectInspector`:
- building: CheckoutBuggyBuild + BuildProject
- testing: TestProject

When `ProjectInspector4Bears` for failing/passing:
- building: CheckoutBuggyBuild + BuildProject
- testing: TestProject

When `ProjectInspector4Bears` for passing/passing:
- building: CheckoutPatchedBuild + ComputeSourceDir + ComputeTestDir + CheckoutBuggyBuildSourceCode + BuildProject
- testing: TestProject

For Bears, we have two builds in the json file, but I guess that what you thought is only related to the buggy build. Thus I suggest to rename `stepDurations` to `stepDurationsBuggyBuild` to be clear that the durations are only about the buggy build.

fermadeiral commented:
@surli, one thing more: I just saw you put `stepDurations` inside `projectMetrics`. What do you think about creating `reproductionBuggyBuild` in the root and putting inside it `stepDurations` and `reproductionDate`?

surli commented:
So I change the schema the way you're thinking but I also change some things:
1. I distinguished begin and end date for the process
2. I added a total duration of the reproduction
3. I added an array for the duration of each sub-steps: the idea is that this array is ordered the same way than the array of steps names. Then it's possible to recover the specific duration of a step from it.

fermadeiral commented:
Great, thanks.
